library/core/src/ptr/mod.rs

```diff
@@ -430,6 +430,10 @@ mod mut_ptr;
 /// done automatically by the compiler. This means the fields of packed structs
 /// are not dropped in-place.
 ///
+/// [`drop_in_place()`] does not modify the pointed-to value beyond any changes
+/// performed by [`Drop::drop()`]. As far as the compiler is concerned, the value
```

Comment:
What is the point of this guarantee? `drop` can still mutate the value pretty much arbitrarily, right?
Or is it intended to say something about the auto-generated drop glue, in case I control the `drop` being called? That seems odd to state like this though, instead we should give a somewhat operational description of the auto-generated drop glue:
- for a struct, it will first call the user-defined `drop` if it exists, and then call the drop glue of each field
- for an enum, it will first call the user-defined `drop` if it exists, and then determine the active variant and call the drop glue for all fields of that variant
- for an array, it will call the drop glue for each field
- ...

Also IMO we shouldn't duplicate the same thing between ManuallyDrop::drop and  here. `drop_in_place` is probably the more canonical location, but then ManuallyDrop::drop should just reference this.

Reply:
> Or is it intended to say something about the auto-generated drop glue, in case I control the drop being called? 

Yes.